src/include/api/vamana_index.h

```diff
@@ -0,0 +1,498 @@
+     *
+     *   @todo Unify the type-based switch-case statements in a manner
+     *   similar to what was done in query_condition
+     */
+    if (feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<uint8_t, uint32_t, uint32_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<float, uint32_t, uint32_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<uint8_t, uint32_t, uint64_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT32 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<float, uint32_t, uint64_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<uint8_t, uint64_t, uint32_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT32) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<float, uint64_t, uint32_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_UINT8 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<uint8_t, uint64_t, uint64_t>>>(
+          ctx, group_uri);
+    } else if (
+        feature_datatype_ == TILEDB_FLOAT32 && id_datatype_ == TILEDB_UINT64 &&
+        px_datatype_ == TILEDB_UINT64) {
+      index_ = std::make_unique<
+          index_impl<detail::graph::vamana_index<float, uint64_t, uint64_t>>>(
+          ctx, group_uri);
+    } else {
+      throw std::runtime_error("Unsupported datatype combination");
+    }
+    if (dimension_ != 0 && dimension_ != index_->dimension()) {
+      throw std::runtime_error(
+          "Dimension mismatch: " + std::to_string(dimension_) +
+          " != " + std::to_string(index_->dimension()));
+    }
+    dimension_ = index_->dimension();
+  }
+
+  /**
+   * @brief Train the index based on the given training set.
+   * @param training_set
+   * @param init
+   */
+  // @todo -- infer feature type from input
+  void train(const FeatureVectorArray& training_set) {
```

Comment:
For a future PR, swap `train` and `add` per our discussion today.  The C++ implementation will need to be swapped as well.  (Or we could go ahead and do it for this PR.)